packages/services/src/contents/index.ts

```diff
@@ -1571,3 +1582,235 @@ namespace Private {
+      throw err;
+    }
+    const data = await response.json();
+    validate.validateContentsModel(data);
+    this.fileChanged.emit({
+      type: 'save',
+      oldValue: null,
+      newValue: data
+    });
+    return data;
+  }
+
+  /**
+   * A signal emitted when a file operation takes place.
+   */
+  get fileChanged(): Signal<this, Contents.IChangedArgs> {
+    return this._options.fileChanged as unknown as Signal<
+      this,
+      Contents.IChangedArgs
+    >;
+  }
+
+  /**
+   * Get a REST url for a file given a path.
+   */
+  private _getUrl(...args: string[]): string {
+    const parts = args.map(path => URLExt.encodeParts(path));
+    const baseUrl = this._options.serverSettings.baseUrl;
+    return URLExt.join(baseUrl, this._options.apiEndpoint, ...parts);
+  }
+
+  private _options: RestContentProvider.IOptions;
+}
+
+namespace RestContentProvider {
+  export interface IOptions {
+    apiEndpoint: string;
+    serverSettings: ServerConnection.ISettings;
+    fileChanged: ISignal<Contents.IDrive, Contents.IChangedArgs>;
+  }
+}
+
+/**
+ * Registry of content providers.
+ * @experimental
+ */
+export interface IContentProviderRegistry {
+  /**
+   * Add a content provider into the registry.
+   *
+   * @param provider - The content provider to register.
+   */
+  register(provider: IContentProvider): void;
+
+  /**
+   * Get a content provider for the given file path.
+   *
+   * @param filePath - The file path for which to get a content provider.
+   */
+  getProvider(filePath: string): IContentProvider | undefined;
+}
+
+/**
+ * The content provider interface.
+ * @experimental
+ */
+export interface IContentProvider {
+  /**
+   * An optional shared model factory instance for the content provider.
+   */
+  readonly sharedModelFactory?: Contents.ISharedFactory;
+
+  /**
+   * The file extensions that this provider handles.
+   */
+  readonly extensions: IContentProviderExtension[];
+
+  /**
+   * Get the file content.
+   *
+   * @param localPath - The path to the file.
+   *
+   * @param options - The optional options for fetching the content.
+   *
+   * @returns A promise which resolves with the file content model.
+   */
+  get(
+    localPath: string,
+    options?: Contents.IFetchOptions
+  ): Promise<Contents.IModel>;
+
+  /**
+   * Save a file.
+   *
+   * @param localPath - The desired file path.
+   *
+   * @param options - Optional overrides to the model.
+   *
+   * @returns A promise which resolves with the file content model when the file is saved.
+   */
+  save(
+    localPath: string,
+    options: Partial<Contents.IModel>
+  ): Promise<Contents.IModel>;
+
+  /**
+   * A signal emitted when a file operation takes place.
+   */
+  readonly fileChanged: ISignal<this, Contents.IChangedArgs>;
+}
+
+/**
+ * The content provider extension interface.
+ */
+export interface IContentProviderExtension {
+  /**
+   * The regular expression matching the file extension
+   */
+  re: string;
```

Comment:
Actually I'm wondering if the job of identifying the file type is not already done by the time the content provider is called. We receive `IFetchOptions` that have a `ContentType` and a `FileFormat`, which should identify the MIME type? So maybe the content provider should only advertise the MIME types it is able to handle. And when users choose "Open with ...", they explicitly choose a MIME type, which means they can choose a particular content provider, not just relying on the one which has the best "score".

Reply:
1. Agree to rename `re` → `pattern`
2. Not sure if we should special-case `extensions[]`, it would complicate the code - need to think about it a bit more
3. Agree that we should consider what happens if user chooses a different viewer.  Strictly users choose a viewer (widget factory) rather than anything related to the API or MIME type (strictly the MIME type stays the same). So we go from `WidgetFactory → ModelFactory → {contentType, fileFormat}`.

- `ContentType` is just `file`/`notebook`/`directory`
- `FileFormat` is `'json' | 'text' | 'base64'`

I'm thinking that when I have my special CSV viewer that can access small chunks of 1TB CSV file, I actually don't care if the extension is `.csv`, `.tsv` or different; all I care about is if user chooses my smart CSV viewer, they get rerouted to the content provider which will stream the contents in chunks. We could extend [`IModelFactory`](https://github.com/jupyterlab/jupyterlab/blob/058d05f720447737274dbec9377af7f6c6c3e7fa/packages/docregistry/src/registry.ts#L1173) so that it can pass some kind of a token saying "Use content provider X" (if available).

I imagine that all smart viewers to work with existing constraints would effectively return `json` with description of the file (say how many rows, how many columns), content hash (to allow logic detecting changes to trigger), and instructions for the widget how to request individual data chunks.

Reply:
Maybe we should remove pattern matching altogether and explicitly support two use cases:
- a) register a new content provider for use with all files (without having control over the widgets)
- b) register a new content provider for use with a widget which request it (in an extension having control over that widget)

We may want to allow paramterizing the `Contents.IModel` returned by the content provider for type safety.

(a) is the RTC use case.

Reply:
The use case of "Open with" was also independently brought up by @bollwyvl.

Reply:
Examples: open a file inside a PDF (because PDFs are actually a folder).

> Personally I received research data this way.